File: Source/UI/Frames/RewardsFrame.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using WCSharp.Api;
using static WCSharp.Api.Common;

public static class RewardsFrame
{
    public static framehandle RewardFrame;
    private static framehandle GameUI = originframetype.GameUI.GetOriginFrame(0);
    private static framehandle TempHandle;
    private static Dictionary<string, framehandle> FrameByName = new Dictionary<string, framehandle>();
    private static Dictionary<framehandle, Reward> RewardIcons = new Dictionary<framehandle, Reward>();
    private static RewardHelper RewardHelp = new RewardHelper();
    private static int RewardsPerRow = 6;
    private static float FrameX = 0.4f;
    private static float FrameY = 0.35f;
    private static float FrameWidth = 0.30f;
    private static float FrameHeight = 0.29f;
    private static float Padding = 0.01f;
    private static float IconSize = 0.02f;
    private static int FrameCount = 0;

    public static void Initialize()
    {
        try
        {
            RewardFrame = CreateRewardFrame();
            TempHandle = RewardFrame;
            SetRewardsFrameHotkey();
            CountRewardFrames();
            AppendRewardsToFrames();
            CreateRandomRewardButton();
            FrameManager.CreateHeaderFrame(RewardFrame);
        }
        catch (Exception ex)
        {
            Logger.Critical($"Error in RewardsFrame: {ex.Message}");
            throw;
        }
    }

    private static framehandle CreateRewardFrame()
    {
        RewardFrame = framehandle.Create("BACKDROP", "Reward Frame", GameUI, "QuestButtonPushedBackdropTemplate", 0);
        RewardFrame.SetAbsPoint(framepointtype.Center, FrameX, FrameY);
        RewardFrame.SetSize(FrameWidth, FrameHeight);

        RewardFrame.Visible = false;
        return RewardFrame;
    }

    private static void CountRewardFrames()
    {
        var count = 0;
        var colCount = 0;
        var rewardTypes = Enum.GetValues(typeof(RewardType)).Cast<RewardType>().OrderBy(rt => (int)rt).ToList(); // sorts by enum, no desync possibilies here.

        foreach (var type in rewardTypes)
        {
            var numberOfRewards = CountNumberOfRewards(type);
            if (numberOfRewards == 0) continue;

            colCount = numberOfRewards / RewardsPerRow;
            if (numberOfRewards % RewardsPerRow != 0) colCount++;
            if (colCount == 0) colCount = 1;

            // Create panel based on number of columns
            InitializePanels(Enum.GetName(typeof(RewardType), type), colCount);
            count += colCount;
            FrameCount++;
        }
        //Console.WriteLine($"There are {count} types");
    }

    private static int CountNumberOfRewards(RewardType type)
    {
        var count = 0;
        foreach (var reward in RewardsManager.Rewards)
        {
            if (reward.Type == type) count++;
        }
        return count;
    }

    private static void InitializePanels(string name, float colCount)
    {
        var y = -Padding;
        var width = (FrameWidth - (Padding * 2)) / 2;
        var height = (colCount == 1) ? 0.03f : 0.03f + ((colCount - 1) * 0.02f);
        CreatePanel(TempHandle, 0, y, width, height, name);
    }

    private static framehandle CreatePanel(framehandle parent, float x, float y, float width, float height, string title)
    {
        var framePoint1 = framepointtype.TopLeft;
        var framePoint2 = framepointtype.BottomLeft;
        if (parent == RewardFrame) framePoint2 = framepointtype.TopLeft;
        if (parent == RewardFrame) x = Padding;
        if (parent == RewardFrame) y = -Padding * 2;
        if (FrameCount == 5)
        {
            parent = RewardFrame;
            framePoint1 = framepointtype.TopRight;
            framePoint2 = framepointtype.TopRight;
            x = -Padding;
            y = -Padding * 2;
        }

        var panel = framehandle.Create("BACKDROP", $"{title}", parent, "QuestButtonDisabledBackdropTemplate", 0);
        panel.SetPoint(framePoint1, x, y, parent, framePoint2);
        panel.SetSize(width, height);

        // title above each panel
        var panelTitle = framehandle.Create("TEXT", "RewardPanelTitle", panel, "", 0);
        panelTitle.SetPoint(framepointtype.TopLeft, 0, Padding, panel, framepointtype.TopLeft);
        panelTitle.Text = title;

        TempHandle = panel;

        FrameByName.Add(title, panel);

        return panel;
    }

    private static void CreateRandomRewardButton()
    {
        // Create the random rewards button at the bottom right of the rewards frame.
        const float ButtonSize = 0.025f;
        const float TooltipWidth = 0.25f;
        const float BackgroundPadding = 0.01f;
        const float TooltipYOffset = 0.01f;
        var dice = "ReplaceableTextures\\CommandButtons\\BTNDice.blp";

        var button = framehandle.Create("Button", "RandomRewardButton", RewardFrame, "ScoreScreenTabButtonTemplate", 0);
        button.SetPoint(framepointtype.BottomRight, -Padding, Padding, RewardFrame, framepointtype.BottomRight);
        button.SetSize(ButtonSize, ButtonSize);

        var icon = framehandle.Create("BACKDROP", "RandomRewardButtonIcon", button, "", 0);
        icon.SetPoints(button);

        var background = framehandle.Create("QuestButtonBaseTemplate", GameUI, 0, 0);
        var tooltipText = framehandle.Create("TEXT", "RandomRewardTooltip", background, "", 0);
        tooltipText.SetSize(TooltipWidth, 0);
        background.SetPoint(framepointtype.BottomLeft, -BackgroundPadding, -BackgroundPadding, tooltipText, framepointtype.BottomLeft);
        background.SetPoint(framepointtype.TopRight, BackgroundPadding, BackgroundPadding, tooltipText, framepointtype.TopRight);

        button.SetTooltip(background);
        tooltipText.SetPoint(framepointtype.Bottom, 0, TooltipYOffset, button, framepointtype.Top);
        tooltipText.Enabled = false;

        icon.SetTexture(dice, 0, false);

        tooltipText.Text = $"{Colors.COLOR_YELLOW}Randomize Rewards{Colors.COLOR_RESET}\n{Colors.COLOR_ORANGE}Picks from your rewards list, applying random cosmetics.{Colors.COLOR_RESET}";

        var t = trigger.Create();
        t.RegisterFrameEvent(button, frameeventtype.Click);
        t.AddAction(RandomRewardsButtonActions);
    }

    private static void RandomRewardsButtonActions()
    {
        var player = @event.Player;
        var frame = @event.Frame;
        try
        {
            RewardHelp.ClearRewards();

            foreach (var reward in RewardsManager.Rewards)
            {
                var stats = Globals.ALL_KITTIES[player].SaveData;
                var value = RewardHelper.GetAwardNestedValue(stats.GameAwardsSorted, reward.TypeSorted, reward.Name);
                if (value <= 0) continue;

                RewardHelp.AddReward(reward);
            }

            Reward selectedHat = (RewardHelp.Hats.Count > 0) ? RewardHelp.Hats[GetRandomInt(0, RewardHelp.Hats.Count - 1)] : null;
            Reward selectedWings = (RewardHelp.Wings.Count > 0) ? RewardHelp.Wings[GetRandomInt(0, RewardHelp.Wings.Count - 1)] : null;
            Reward selectedTrail = (RewardHelp.Trails.Count > 0) ? RewardHelp.Trails[GetRandomInt(0, RewardHelp.Trails.Count - 1)] : null;
            Reward selectedAura = (RewardHelp.Auras.Count > 0) ? RewardHelp.Auras[GetRandomInt(0, RewardHelp.Auras.Count - 1)] : null;

            selectedHat?.ApplyReward(player);
            selectedWings?.ApplyReward(player);
            selectedTrail?.ApplyReward(player);
            selectedAura?.ApplyReward(player);

            if (!player.IsLocal) return;
            FrameManager.RefreshFrame(frame);
        }
        catch (Exception e)
        {
            Logger.Warning($"Error in RandomRewardsButtonActions: {e.Message}");
        }
    }

    private static void AppendRewardsToFrames()
    {
        var cols = RewardsPerRow;
        var count = new Dictionary<RewardType, int>();

        // Initialize the count dictionary with zeros
        foreach (RewardType type in Enum.GetValues(typeof(RewardType)))
        {
            count[type] = 0;
        }

        foreach (var reward in RewardsManager.Rewards)
        {
            count[reward.Type]++;
            var rows = (CountNumberOfRewards(reward.Type) / cols) + 1;
            var col = (count[reward.Type] - 1) % cols;
            var row = (count[reward.Type] - 1) / cols;
            if (!FrameByName.TryGetValue(Enum.GetName(typeof(RewardType), reward.Type), out var panel))
            {
                continue;
            }
            var rewardButton = framehandle.Create("Button", reward.Name.ToString(), panel, "ScoreScreenTabButtonTemplate", 0);
            if (col == 0)
                rewardButton.SetPoint(framepointtype.TopLeft, Padding, (-row * IconSize) - (Padding / 2), panel, framepointtype.TopLeft);
            else
                rewardButton.SetPoint(framepointtype.TopLeft, (col * IconSize) + Padding, (-row * IconSize) - (Padding / 2), panel, framepointtype.TopLeft);
            rewardButton.SetSize(IconSize, IconSize);

            var icon = framehandle.Create("BACKDROP", reward.Name.ToString() + "icon", rewardButton, "", 0);
            icon.SetPoints(rewardButton);
            RewardTooltip(rewardButton, reward);

            RewardIcons.Add(icon, reward);

            var Trigger = trigger.Create();
            Trigger.RegisterFrameEvent(rewardButton, frameeventtype.Click);
            Trigger.AddAction(() => RewardButtonActions(reward));
        }

        GC.RemoveDictionary(ref count);
    }

    private static void RewardTooltip(framehandle parent, Reward reward)
    {
        var background = framehandle.Create("QuestButtonBaseTemplate", GameUI, 0, 0);
        var tooltipText = framehandle.Create("TEXT", $"{reward.Name}Tooltip", background, "", 0);

        tooltipText.SetSize(0.25f, 0);
        background.SetPoint(framepointtype.BottomLeft, -0.01f, -0.01f, tooltipText, framepointtype.BottomLeft);
        background.SetPoint(framepointtype.TopRight, 0.01f, 0.01f, tooltipText, framepointtype.TopRight);

        parent.SetTooltip(background);
        tooltipText.SetPoint(framepointtype.Bottom, 0, 0.01f, parent, framepointtype.Top);
        tooltipText.Enabled = false;

        var name = BlzGetAbilityTooltip(reward.AbilityID, 0);
        var desc = BlzGetAbilityExtendedTooltip(reward.AbilityID, 0);

        tooltipText.Text = $"{name}\n{desc}";
    }

    private static void RewardButtonActions(Reward reward)
    {
        var player = @event.Player;
        var frame = @event.Frame;
        var stats = Globals.ALL_KITTIES[player].SaveData;
        var value = RewardHelper.GetAwardNestedValue(stats.GameAwardsSorted, reward.TypeSorted, reward.Name);
        if (value <= 0) return; // Doesnt have the reward , dont apply.
        reward.ApplyReward(player);
        if (!player.IsLocal) return;
        FrameManager.RefreshFrame(frame);
    }

    private static void UnavilableRewardIcons(player player)
    {
        var stats = Globals.ALL_KITTIES[player].SaveData;
        var unavailablePath = "ReplaceableTextures\\CommandButtons\\BTNSelectHeroOn";
        foreach (var reward in RewardIcons)
        {
            var value = RewardHelper.GetAwardNestedValue(stats.GameAwardsSorted, reward.Value.TypeSorted, reward.Value.Name);
            if (value <= 0) // Doesnt have reward
                reward.Key.SetTexture(unavailablePath, 0, false);
            else
                reward.Key.SetTexture(BlzGetAbilityIcon(reward.Value.AbilityID), 0, false);
        }
    }

    private static void SetRewardsFrameHotkey()
    {
        var rewardsHotKey = trigger.Create();
        foreach (var player in Globals.ALL_PLAYERS)
        {
            rewardsHotKey.RegisterPlayerKeyEvent(player, OSKEY_OEM_MINUS, 0, true);
        }
        rewardsHotKey.AddAction(ErrorHandler.Wrap(() =>
        {
            RewardsFrameActions();
        }));
    }

    public static void RewardsFrameActions()
    {
        var player = @event.Player;
        if (!player.IsLocal) return;
        // if (ShopUtil.IsPlayerInWolfLane(player)) return;
        FrameManager.RewardsButton.Visible = false;
        FrameManager.RewardsButton.Visible = true;
        FrameManager.HideOtherFrames(RewardFrame);
        RewardFrame.Visible = !RewardFrame.Visible;
        if (RewardFrame.Visible)
        {
            UnavilableRewardIcons(player);
            MultiboardUtil.MinMultiboards(player, true);
        }
    }
}

File: src/UI/Frames/RewardsFrame.ts
import { Logger } from 'src/Events/Logger/Logger'
import { CurrentGameMode } from 'src/Gamemodes/CurrentGameMode'
import { GameMode } from 'src/Gamemodes/GameModeEnum'
import { Globals } from 'src/Global/Globals'
import { Reward, RewardType } from 'src/Rewards/Rewards/Reward'
import { RewardsManager } from 'src/Rewards/Rewards/RewardsManager'
import { Colors } from 'src/Utility/Colors/Colors'
import { blzCreateFrame, blzCreateFrameByType, getTriggerPlayer } from 'src/Utility/w3tsUtils'
import { Frame, MapPlayer, Trigger } from 'w3ts'
import { MultiboardUtil } from '../Multiboard/MultiboardUtil'
import { FrameManager } from './FrameManager'
import { CreateHeaderFrame, HideOtherFrames } from './FrameUtil'
import { RewardHelper } from './RewardHelper'

export class RewardsFrame {
    public static RewardFrame: Frame
    private static GameUI: Frame
    private static TempHandle: Frame
    private static FrameByName: Map<string, Frame> = new Map()
    private static RewardIcons: Map<Frame, Reward> = new Map()
    private static RewardHelp: RewardHelper = new RewardHelper()
    private static RewardsPerRow = 6
    private static FrameX = 0.4
    private static FrameY = 0.35
    private static FrameWidth = 0.3
    private static FrameHeight = 0.29
    private static Padding = 0.01
    private static IconSize = 0.02
    private static FrameCount = 0

    public static Initialize = () => {
        try {
            RewardsFrame.GameUI = Frame.fromHandle(BlzGetOriginFrame(ORIGIN_FRAME_GAME_UI, 0))!
            RewardsFrame.RewardFrame = RewardsFrame.CreateRewardFrame()
            RewardsFrame.TempHandle = RewardsFrame.RewardFrame
            RewardsFrame.SetRewardsFrameHotkey()
            RewardsFrame.CountRewardFrames()
            RewardsFrame.AppendRewardsToFrames()
            RewardsFrame.CreateRandomRewardButton()
            CreateHeaderFrame(RewardsFrame.RewardFrame)
        } catch (ex: any) {
            Logger.Critical(`Error in RewardsFrame: ${ex}`)
            throw ex
        }
    }

    private static CreateRewardFrame(): Frame {
        RewardsFrame.RewardFrame = blzCreateFrameByType(
            'BACKDROP',
            'Frame: Reward',
            RewardsFrame.GameUI,
            'QuestButtonPushedBackdropTemplate',
            0
        )!
        RewardsFrame.RewardFrame.setAbsPoint(FRAMEPOINT_CENTER, RewardsFrame.FrameX, RewardsFrame.FrameY)
        RewardsFrame.RewardFrame.setSize(RewardsFrame.FrameWidth, RewardsFrame.FrameHeight)

        RewardsFrame.RewardFrame.visible = false
        return RewardsFrame.RewardFrame
    }

    private static CountRewardFrames = () => {
        let count = 0
        let colCount = 0
        const rewardTypes = Object.values(RewardType)
            .filter((v): v is RewardType => typeof v === 'number')
            .sort((a, b) => a - b)

        // sorts by enum, no desync possibilies here.

        for (const type of rewardTypes) {
            const numberOfRewards = RewardsFrame.CountNumberOfRewards(type)
            if (numberOfRewards === 0) continue

            colCount = numberOfRewards / RewardsFrame.RewardsPerRow
            if (numberOfRewards % RewardsFrame.RewardsPerRow !== 0) colCount++
            if (colCount === 0) colCount = 1

            // Create panel based on number of columns
            RewardsFrame.InitializePanels(RewardType[type], colCount)
            count += colCount
            RewardsFrame.FrameCount++
        }
        //print(`There are ${count} types`);
    }

    private static CountNumberOfRewards(type: RewardType) {
        let count = 0
        for (const reward of RewardsManager.Rewards) {
            if (reward.Type === type) count++
        }
        return count
    }

    private static InitializePanels(name: string, colCount: number) {
        const y = -RewardsFrame.Padding
        const width = (RewardsFrame.FrameWidth - RewardsFrame.Padding * 2) / 2
        const height = colCount === 1 ? 0.03 : 0.03 + (colCount - 1) * 0.02
        RewardsFrame.CreatePanel(RewardsFrame.TempHandle, 0, y, width, height, name)
    }

    private static CreatePanel(
        parent: Frame,
        x: number,
        y: number,
        width: number,
        height: number,
        title: string
    ): Frame {
        let framePoint1 = FRAMEPOINT_TOPLEFT
        let framePoint2 = FRAMEPOINT_BOTTOMLEFT
        if (parent === RewardsFrame.RewardFrame) framePoint2 = FRAMEPOINT_TOPLEFT
        if (parent === RewardsFrame.RewardFrame) x = RewardsFrame.Padding
        if (parent === RewardsFrame.RewardFrame) y = -RewardsFrame.Padding * 2
        if (RewardsFrame.FrameCount === 5) {
            parent = RewardsFrame.RewardFrame
            framePoint1 = FRAMEPOINT_TOPRIGHT
            framePoint2 = FRAMEPOINT_TOPRIGHT
            x = -RewardsFrame.Padding
            y = -RewardsFrame.Padding * 2
        }

        const panel = blzCreateFrameByType('BACKDROP', `${title}`, parent, 'QuestButtonDisabledBackdropTemplate', 0)
        panel.setPoint(framePoint1, parent, framePoint2, x, y)
        panel.setSize(width, height)

        // title above each panel
        const panelTitle = blzCreateFrameByType('TEXT', 'RewardPanelTitle', panel, '', 0)
        panelTitle.setPoint(FRAMEPOINT_TOPLEFT, panel, FRAMEPOINT_TOPLEFT, 0, RewardsFrame.Padding)
        panelTitle.text = title

        RewardsFrame.TempHandle = panel

        RewardsFrame.FrameByName.set(title, panel)

        return panel
    }

    private static CreateRandomRewardButton = () => {
        // Create the random rewards button at the bottom right of the rewards frame.
        const ButtonSize = 0.025
        const TooltipWidth = 0.25
        const BackgroundPadding = 0.01
        const TooltipYOffset = 0.01
        const dice = 'ReplaceableTextures\\CommandButtons\\BTNDice.blp'

        const button = blzCreateFrameByType(
            'Button',
            'RandomRewardButton',
            RewardsFrame.RewardFrame,
            'ScoreScreenTabButtonTemplate',
            0
        )
        button.setPoint(
            FRAMEPOINT_BOTTOMRIGHT,
            RewardsFrame.RewardFrame,
            FRAMEPOINT_BOTTOMRIGHT,
            -RewardsFrame.Padding,
            RewardsFrame.Padding
        )
        button.setSize(ButtonSize, ButtonSize)

        const icon = blzCreateFrameByType('BACKDROP', 'RandomRewardButtonIcon', button, '', 0)
        icon.setAllPoints(button)

        const background = blzCreateFrame('QuestButtonBaseTemplate', RewardsFrame.GameUI, 0, 0)
        const tooltipText = blzCreateFrameByType('TEXT', 'RandomRewardTooltip', background, '', 0)
        tooltipText.setSize(TooltipWidth, 0)
        background.setPoint(
            FRAMEPOINT_BOTTOMLEFT,
            tooltipText,
            FRAMEPOINT_BOTTOMLEFT,
            -BackgroundPadding,
            -BackgroundPadding
        )
        background.setPoint(FRAMEPOINT_TOPRIGHT, tooltipText, FRAMEPOINT_TOPRIGHT, BackgroundPadding, BackgroundPadding)

        button.setTooltip(background)
        tooltipText.setPoint(FRAMEPOINT_BOTTOM, button, FRAMEPOINT_TOP, 0, TooltipYOffset)
        tooltipText.enabled = false

        icon.setTexture(dice, 0, false)

        tooltipText.text = `${Colors.COLOR_YELLOW}Randomize Rewards${Colors.COLOR_RESET}\n${Colors.COLOR_ORANGE}Picks random cosmetics from your rewards list, applying them.${Colors.COLOR_RESET}`

        const t = Trigger.create()!
        t.triggerRegisterFrameEvent(button, FRAMEEVENT_CONTROL_CLICK)
        t.addAction(RewardsFrame.RandomRewardsButtonActions)
    }

    private static RandomRewardsButtonActions = () => {
        const player = getTriggerPlayer()
        const frame = Frame.fromHandle(BlzGetTriggerFrame())!
        try {
            RewardsFrame.RewardHelp.ClearRewards()

            for (const reward of RewardsManager.Rewards) {
                const stats = Globals.ALL_KITTIES.get(player)!.SaveData
                const value = RewardHelper.GetAwardNestedValue(stats.GameAwardsSorted, reward.TypeSorted, reward.name)
                if (value <= 0) continue

                RewardsFrame.RewardHelp.AddReward(reward)
            }

            const selectedHat =
                RewardsFrame.RewardHelp.Hats.length > 0
                    ? RewardsFrame.RewardHelp.Hats[GetRandomInt(0, RewardsFrame.RewardHelp.Hats.length - 1)]
                    : null
            const selectedWings =
                RewardsFrame.RewardHelp.Wings.length > 0
                    ? RewardsFrame.RewardHelp.Wings[GetRandomInt(0, RewardsFrame.RewardHelp.Wings.length - 1)]
                    : null
            const selectedTrail =
                RewardsFrame.RewardHelp.Trails.length > 0
                    ? RewardsFrame.RewardHelp.Trails[GetRandomInt(0, RewardsFrame.RewardHelp.Trails.length - 1)]
                    : null
            const selectedAura =
                RewardsFrame.RewardHelp.Auras.length > 0
                    ? RewardsFrame.RewardHelp.Auras[GetRandomInt(0, RewardsFrame.RewardHelp.Auras.length - 1)]
                    : null

            selectedHat?.ApplyReward(player)
            selectedWings?.ApplyReward(player)
            selectedTrail?.ApplyReward(player)
            selectedAura?.ApplyReward(player)

            if (!player.isLocal()) return
            FrameManager.RefreshFrame(frame)
        } catch (e: any) {
            Logger.Warning(`Error in RandomRewardsButtonActions: ${e}`)
        }
    }

    private static AppendRewardsToFrames = () => {
        const cols = RewardsFrame.RewardsPerRow
        const count: Map<number, number> = new Map()

        for (const reward of RewardsManager.Rewards) {
            const type = reward.Type
            count.set(type, (count.get(type) || 0) + 1)

            const totalRewards = RewardsFrame.CountNumberOfRewards(reward.Type)
            const col = ((count.get(type) || 0) - 1) % cols
            const row = Math.floor(((count.get(type) || 0) - 1) / cols)

            // Get the panel for RewardsFrame reward type
            const panelName = RewardType[reward.Type]
            const panel = RewardsFrame.FrameByName.get(panelName)
            if (!panel) continue

            // Create the reward button
            const rewardButton = blzCreateFrameByType(
                'Button',
                reward.name.toString(),
                panel,
                'ScoreScreenTabButtonTemplate',
                0
            )

            const x = col * RewardsFrame.IconSize + RewardsFrame.Padding
            const y = -row * RewardsFrame.IconSize - RewardsFrame.Padding / 2
            rewardButton.setPoint(FRAMEPOINT_TOPLEFT, panel, FRAMEPOINT_TOPLEFT, x, y)
            rewardButton.setSize(RewardsFrame.IconSize, RewardsFrame.IconSize)

            // Create the icon
            const icon = blzCreateFrameByType('BACKDROP', reward.name.toString() + 'icon', rewardButton, '', 0)
            icon.setAllPoints(rewardButton)

            RewardsFrame.RewardTooltip(rewardButton, reward)
            RewardsFrame.RewardIcons.set(icon, reward)

            // Register click event
            const triggerObj = Trigger.create()!
            triggerObj.triggerRegisterFrameEvent(rewardButton, FRAMEEVENT_CONTROL_CLICK)
            triggerObj.addAction(() => RewardsFrame.RewardButtonActions(reward))
        }
    }

    private static RewardTooltip(parent: Frame, reward: Reward) {
        const background = blzCreateFrame('QuestButtonBaseTemplate', RewardsFrame.GameUI, 0, 0)
        const tooltipText = blzCreateFrameByType('TEXT', `${reward.name}Tooltip`, background, '', 0)

        tooltipText.setSize(0.25, 0)
        background.setPoint(FRAMEPOINT_BOTTOMLEFT, tooltipText, FRAMEPOINT_BOTTOMLEFT, -0.01, -0.01)
        background.setPoint(FRAMEPOINT_TOPRIGHT, tooltipText, FRAMEPOINT_TOPRIGHT, 0.01, 0.01)

        parent.setTooltip(background)
        tooltipText.setPoint(FRAMEPOINT_BOTTOM, parent, FRAMEPOINT_TOP, 0, 0.01)
        tooltipText.enabled = false

        const name = BlzGetAbilityTooltip(reward.AbilityID, 0)
        const desc = BlzGetAbilityExtendedTooltip(reward.AbilityID, 0)

        tooltipText.text = `${name}\n${desc}`
    }

    private static RewardButtonActions(reward: Reward) {
        const player = getTriggerPlayer()
        const frame = Frame.fromEvent()
        const stats = Globals.ALL_KITTIES.get(player)!.SaveData
        const value = RewardHelper.GetAwardNestedValue(stats.GameAwardsSorted, reward.TypeSorted, reward.name)
        if (value <= 0) return // Doesnt have the reward , dont apply.
        reward.ApplyReward(player)
        if (!player.isLocal()) return
        if (!frame) return
        FrameManager.RefreshFrame(frame)
    }

    private static UnavilableRewardIcons(player: MapPlayer) {
        try {
            const stats = Globals.ALL_KITTIES.get(player)!.SaveData
            const unavailablePath = 'ReplaceableTextures\\CommandButtons\\BTNSelectHeroOn'
            for (const [frame, reward] of RewardsFrame.RewardIcons) {
                const value = RewardHelper.GetAwardNestedValue(stats.GameAwardsSorted, reward.TypeSorted, reward.name)
                if (value <= 0)
                    // Doesnt have reward
                    frame.setTexture(unavailablePath, 0, false)
                else frame.setTexture(BlzGetAbilityIcon(reward.AbilityID)!, 0, false)
            }
        } catch (e: any) {
            Logger.Warning(`Error in UnavilableRewardIcons: ${e}`)
        }
    }

    private static SetRewardsFrameHotkey = () => {
        const rewardsHotKey = Trigger.create()!
        for (const player of Globals.ALL_PLAYERS) {
            rewardsHotKey.registerPlayerKeyEvent(player, OSKEY_OEM_MINUS, 0, true)
        }
        rewardsHotKey.addAction(RewardsFrame.RewardsFrameActions)
    }

    public static RewardsFrameActions = () => {
        const player = getTriggerPlayer()
        if (!player.isLocal()) return
        if (CurrentGameMode.active !== GameMode.Standard) {
            player.DisplayTimedTextTo(
                3.0,
                `${Colors.COLOR_RED}Rewards are only available in Standard Mode${Colors.COLOR_RESET}`
            )
            return // Let's not activate rewards in tournament.
        }
        // if (ShopUtil.IsPlayerInWolfLane(player)) return;
        FrameManager.RewardsButton.visible = false
        FrameManager.RewardsButton.visible = true
        HideOtherFrames(RewardsFrame.RewardFrame)
        RewardsFrame.RewardFrame.visible = !RewardsFrame.RewardFrame.visible
        if (RewardsFrame.RewardFrame.visible) {
            RewardsFrame.UnavilableRewardIcons(player)
            MultiboardUtil.MinMultiboards(player, true)
        }
    }
}
